### PR TITLE
Fix listenTo memory leak

### PR DESCRIPTION
**Pull request by jridgewell**
Fixes https://github.com/jashkenas/backbone/issues/3453, and competes with https://github.com/jashkenas/backbone/pull/3049.

I think this aligns more closely with what's currently implemented. :smile:

It also implements `listenApi`, which is `eventsApi` translated to the `listenTo` / `stopListening` method signatures.

### CODE REVIEW

File: test/events.js [hunk 1]
```diff
@@ -79,7 +79,7 @@
   test("listenTo and stopListening with event maps", 4, function() {
     var a = _.extend({}, Backbone.Events);
     var b = _.extend({}, Backbone.Events);
-    var cb = function(){ ok(true); };
+    var cb = function(){ console.log('called'); ok(true); };
```
Comment (megawac) [inline test/events.js]: Whoops :)

Comment (jridgewell) [inline test/events.js]: :smile:

**Comment by jridgewell** [comment]
As a side: I have a completed [`eventsApi` refactor](https://github.com/jridgewell/backbone/compare/eventsApi-refactor) that cleans up the duplication in `listenApi`. I haven't included it in this PR yet, since it might be polarizing. If the only issue with this PR is the duplication, I'd be glad to put it in as well.

**Comment by jashkenas** [comment]
Cleaning up the duplication would be lovely.

**Comment by jridgewell** [comment]
Ok, it's included now.

File: backbone.js [hunk 2]
```diff
@@ -224,27 +158,99 @@
...
+      if (events) triggerEvents(events, args);
+      if (allEvents) triggerEvents(allEvents, [name].concat(args));
+    }
+    return obj;
+  };
+
+  // Maps the normalized event callbacks into onceWrappers.
+  var onceMap = function(name, callback, offer) {
+    return eventsApi(function(map, name, callback, offer) {
+      if (callback) map[name] = onceWrap(name, callback, offer);
```
Comment (akre54) [inline backbone.js]: Any reason not to inline `onceWrap` here?

File: backbone.js [hunk 3]
```diff
@@ -205,14 +136,17 @@
...
       if (!listeningTo) return this;
-      var remove = !name && !callback;
-      if (!callback && typeof name === 'object') callback = this;
-      if (obj) (listeningTo = {})[obj._listenId] = obj;
-      for (var id in listeningTo) {
-        obj = listeningTo[id];
-        obj.off(name, callback, this);
-        if (remove || _.isEmpty(obj._events)) delete this._listeningTo[id];
+
+      var listeneeIds = (obj) ? [obj._listenId] : _.keys(listeningTo);
```
Comment (akre54) [inline backbone.js]: unnecessary parens

File: backbone.js [hunk 4]
```diff
@@ -224,27 +158,99 @@
...
       }
-      return false;
     }
+    return _.isEmpty(events) ? void 0 : events;
+  };
 
-    return true;
+  var triggerApi = function(obj, name, sentinel, args) {
+    if (obj._events) {
+      if (sentinel !== triggerApi) args = [sentinel].concat(args);
```
Comment (akre54) [inline backbone.js]: What's the sentinel about? I assume something with the `eventsAPI` works?

Some comments here (and elsewhere) would help.

File: backbone.js (hunk 2 above)
Comment (jridgewell) [inline backbone.js]: Nope. I'll fix it.

File: backbone.js (hunk 4 above)
Comment (jridgewell) [inline backbone.js]: That's because `trigger({ event: value}, 1, 2, 3)` will prepend `value` to the arguments of the listener. Re: https://github.com/jashkenas/backbone/issues/3460

Comment (akre54) [inline backbone.js]: Makes sense. I had a hard time tracking what's going on with the callbacks here, since the only other place I see this called, `triggerApi` is being used as the sentinel.

**Comment by jridgewell** [comment]
Ping @akre54: I commented, and added a `triggerSentinel` object to prevent further confusion. Let me know you have any other suggestions.

File: backbone.js [hunk 5]
```diff
@@ -159,11 +109,9 @@
     trigger: function(name) {
       if (!this._events) return this;
       var args = slice.call(arguments, 1);
-      if (!eventsApi(this, 'trigger', name, args)) return this;
-      var events = this._events[name];
-      var allEvents = this._events.all;
-      if (events) triggerEvents(events, args);
-      if (allEvents) triggerEvents(allEvents, arguments);
+
+      // Use `eventsApi` to normalize `name` into the proper event names.
```
Comment (akre54) [inline backbone.js]: Mind clarifying this? I'm not sure what this is getting at.

**Comment by akre54** [comment]
I think a little more of the "why" and the "how" and less "what" in the comments might be helpful.

Maybe a sentence or two at the top briefly describing *how* the many abstracted functions work. Right now I'm on the fence that this consolidation is worth the added indirection. It definitely makes it harder to follow the logic. 

Can you also make a jsperf showing the extra functions aren't too horrible and squash to a single commit and I'll merge? Thanks for your hard work on this. (And no stress about getting it done today... enjoy the snowday!)

**Comment by jridgewell** [comment]
Here's a simple jsperf: http://jsperf.com/eventsapi-refactor (It was hidden in 50adef3's commit message). Safari seems to be the only one taking a hit, but it's still very fast.

**Comment by akre54** [comment]
how about vs master? That's showing all the changes with the extra
callbacks added in.

**Comment by jridgewell** [comment]
Here's the simple case [against master](http://jsperf.com/eventsapi-refactor/3), and another using [space-separated events](http://jsperf.com/eventsapi-refactor/2).

I did the original tests against the `listenTo-memory-leak-immutable-on-events` branch, because that's the one with the minimum code to fix the memory leak. Comparing against `master` seemed apples-to-oranges.

**Comment by jridgewell** [comment]
Actually, the performance hit seems to be coming from https://github.com/jashkenas/backbone/pull/3463, which is also merged into `eventsApi-refactor`. The [simple case](http://jsperf.com/eventsapi-refactor/4) against those two have the exact same speeds.

**Comment by smelnikov** [comment]
Nice, but does not seem to compete with #3049 as it does not address the following: 

```javascript
var a = _.extend({}, Backbone.Events);
var b = _.extend({}, Backbone.Events);
var c = _.extend({}, Backbone.Events);

a.listenTo(c, 'x', fn);
b.listenTo(c, 'x', fn);
c.off('x');
c = null; //c can't be garbage collected until 'stopListening()' is called on a and b
```

Here are the relevant tests

```javascript
 test("off removes object from listeners", 6, function() {                                                                                                                                                                                    
    var a = _.extend({}, Backbone.Events);
    var b = _.extend({}, Backbone.Events);
    var c = _.extend({}, Backbone.Events);
    var d = _.extend({}, Backbone.Events);
    var fn = function() {};    
        
    b.listenTo(a, 'all', fn);  
    c.listenTo(a, 'all', fn);  
    c.listenTo(d, 'x', fn);    
      
    equal(_.size(b._listeningTo), 1);
    equal(_.size(c._listeningTo), 2);
    equal(_.size(a._listeners), 2); 
    a.off();                   
    equal(_.size(b._listeningTo), 0);
    equal(_.size(c._listeningTo), 1);
    equal(_.size(a._listeners), 0); 
  }); 
      
  test("off w/event name removes object from listeners", 8, function() {
    var a = _.extend({}, Backbone.Events);
    var b = _.extend({}, Backbone.Events);
    var c = _.extend({}, Backbone.Events);
    var fn = function() {};

    b.listenTo(a, 'x', fn);
    c.listenTo(a, 'x y', fn);

    equal(_.size(b._listeningTo), 1);
    equal(_.size(c._listeningTo), 1);
    equal(_.size(a._listeners), 2); 
    a.off('x');
    equal(_.size(b._listeningTo), 0);
    equal(_.size(c._listeningTo), 1);
    equal(_.size(a._listeners), 1); 
    a.off('y');
    equal(_.size(c._listeningTo), 0);
    equal(_.size(a._listeners), 0); 
  });
```

**Comment by jridgewell** [comment]
Hmm, let me see what I can do to support that case.

**Comment by jridgewell** [comment]
I've addressed @smelnikov's `#off` memory leak, but the code isn't as pretty as I would have liked.

File: backbone.js (hunk 5 above)
Comment (akre54) [inline backbone.js]: @jridgewell the comments are still lacking. Can you do a pass to make sure the logic is clear? I'm finding it hard to follow what is happening and why.

Comment (jridgewell) [inline backbone.js]: I haven't revised the comments yet, was trying to solve the `off` problem first.

Is there a particular spot you're not able to follow?

File: backbone.js [hunk 6]
```diff
@@ -106,49 +97,19 @@
...
-          delete this._events[name];
+      if (!this._events) return this;
+      this._events = eventsApi(offApi, this._events, name, callback, context);
+
+      var listeners = this._listeners;
+      if (listeners) {
+        var ids = context != null ? [context._listenId] : _.keys(listeners);
+        for (var i = 0, length = ids.length; i < length; i++) {
+          var listener = listeners[ids[i]];
+          if (!listener) break;
```
Comment (megawac) [inline backbone.js]: Why is this necessary? I would expect a `continue` if anything

File: backbone.js (hunk 5 above)
Comment (akre54) [inline backbone.js]: The bits with `triggerSentinel` and the numerous internal functions mostly. 

It almost seems like we should drop the `Events = {on: ..., off: ..., listenTo: ...}` object, and just colocate the different events methods (`Events.on = ...; onApi = ...; Events.off = ...;`.) To follow this logic, I'm having to jump back and forth through the source to even understand where the different pieces are going.

There's so much extra overhead and indirection going on here, it really detracts from the readability of the Events code.

File: backbone.js [hunk 7]
```diff
@@ -171,48 +130,33 @@
...
-        return this;
-      }
-      if (!callback) return this;
-      var once = _.once(function() {
-        this.stopListening(obj, name, once);
-        callback.apply(this, arguments);
-      });
-      once._callback = callback;
-      return this.listenTo(obj, name, once);
+      name = onceMap(name, callback, _.bind(this.stopListening, this, obj));
```
Comment (megawac) [inline backbone.js]: Should it not also be bound to `name, callback`?

File: backbone.js [hunk 8]
```diff
@@ -223,28 +167,129 @@
...
+    if (name && typeof name === 'object') {
+      // Handle event maps.
+      for (names = _.keys(name), length = names.length; i < length; i++) {
+        events = api(events, names[i], name[names[i]], context, ctx);
       }
-      return false;
+    } else if (name && eventSplitter.test(name)) {
+      // Handle space separated event names.
+      for (names = name.split(eventSplitter), length = names.length; i < length; i++) {
+        events = api(events, names[i], callback, context, ctx);
```
Comment (megawac) [inline backbone.js]: I feel like the redundancy in these two can be addressed but it can be done in a follow up

File: backbone.js (hunk 6 above)
Comment (jridgewell) [inline backbone.js]: When you `listenTo`, context is forced to be `this`. If we're `off`ing, we need to consider two circumstances:

1. No context
    - We'll loop through all our listeners, since any could be affected.
    - This conditional will never be true, so we'll never `break`.
2. Context
    - Only loop through the listeners of the corresponding `_listenId`, since **only** that listener will have the same `context`
    - If we have `listeners`, but not that `_listenId`, then `break`.
      - `ids` is guaranteed to only equal `[context._listenId]`, so there's no use `continue`ing.
    - If we have that `_listenId`, all is good.

File: backbone.js [hunk 9]
```diff
@@ -223,28 +167,129 @@
...
+      events = api(events, name, callback, context, ctx);
+    }
+    return events;
+  };
+
+  // Handles actually adding the event handler.
+  var onApi = function(events, name, callback, context, ctx) {
+    if (callback) {
+      var handlers = events[name] || [];
+      events[name] = handlers.concat({callback: callback, context: context, ctx: context || ctx});
```
Comment (megawac) [inline backbone.js]: What was the reason you didn't want to use push again, I saw the issue but didnt read it?

File: backbone.js (hunk 7 above)
Comment (jridgewell) [inline backbone.js]: Can't, because `name` could be an object or space-separated string. This will be passed in the normalized (just a single event name and it's callback) by the [onceWrapper](https://github.com/jashkenas/backbone/pull/3455/files#diff-0d56d0d310de7ff18b3cef9c2f8f75dcR286).

File: backbone.js [hunk 10]
```diff
@@ -223,28 +167,129 @@
...
+      name = names[i];
 
-    // Handle space separated event names.
-    if (eventSplitter.test(name)) {
-      var names = name.split(eventSplitter);
-      for (var i = 0, length = names.length; i < length; i++) {
-        obj[action].apply(obj, [names[i]].concat(rest));
+      // Bail out if there are no events stored.
+      var handlers = events[name];
+      if (!handlers) break;
```
Comment (megawac) [inline backbone.js]: Again, I think this should be a `continue`

File: backbone.js [hunk 11]
```diff
@@ -223,28 +167,129 @@
...
+              callback !== handler.callback._callback ||
+                context && context !== handler.context
+          ) {
+            remaining.push(handler);
+          }
+        }
       }
-      return false;
+
+      // Replace events if there are any remaining.  Otherwise, clean up.
```
Comment (megawac) [inline backbone.js]: Nit: double space

File: backbone.js (hunk 5 above)
Comment (jridgewell) [inline backbone.js]: I agree the source could be better arranged, I'll work on that tonight along with the comments.

> There's so much extra overhead and indirection going on here

I actually thought it was kind of easier to follow without the old `eventsApi` behavior, but then I had to account for `off` cleaning up any listeners. It makes it easier in my mind to think of all the `*Api` functions as special purpose reduce iteratees, with `eventsApi` being `_.reduce`.

File: backbone.js (hunk 9 above)
Comment (jridgewell) [inline backbone.js]: Agh. I think we've decided to `slice` in `#trigger` instead of `concat` here. I'll need to update this.

File: backbone.js (hunk 10 above)
Comment (jridgewell) [inline backbone.js]: It's the same kind of logic as the other. The only way the conditional will ever be true is if you pass the `name` argument, in which case `names = [name]`. So, no need to continue.

File: backbone.js (hunk 11 above)
Comment (jridgewell) [inline backbone.js]: :stuck_out_tongue:

**Comment by jridgewell** [comment]
@akre54 Commented and rearranged.

**Comment by jashkenas** [comment]
I think let's maybe merge this for now -- and then go over it and tweak as need be. (But hopefully need not ;)
